### ISSUE TITLE
Feature request: show mc during initial selection

### ISSUE THREAD

**Issue by vsrisuknimit** [bafolts/terraforming-mars#1285]
During initial card purchase (after initial drafting is done), once the user select a corporation. Show how much megacredit the user has from that corporation. After click on each prelude and project card, deduct or increase the mc accordingly. This will help users calculate how many cards to keep or what prelude to pick.

**Comment by vsrisuknimit** [comment]
![FireShot Capture 158 - Terraforming Mars - terraforming-mars](https://user-images.githubusercontent.com/14239220/93021918-b56d1f00-f5b3-11ea-889b-06544106ab1c.png)

**Comment by pierrehilbert** [comment]
I will work on it. Should it take into account preludes bonus/malus?
I think it should but would like your advice.

**Comment by vsrisuknimit** [comment]
@pierrehilbert Yes, it should take into account prelude that provide and deduct cash.

**Comment by vsrisuknimit** [comment]
It should also handle Polyphemos and Terralabs differently.

I don't know if it should take into account the extra mc from corp immediate reveal bonus (Vitor and Splice and Pharmacy Union).

**Comment by kevinb9n** [comment]
I think this would be an excellent feature. Some thoughts:

For any MC amount that is guaranteed to be given or deducted by Corp reveal effects, I believe the value *should* account for those adjustments. If it doesn't, I think it will violate user expectations and they will file bugs. And they'd be right to: consider that whether Vitor had been written as "45 MC, effect includes itself" or "48 MC, effect excludes itself" is only a *wording* choice that has no in-game consequence.

The number you calculate might be negative. If so, I would still just display it. It will remind the player to adjust their selections until it is at least zero. (That still doesn't guarantee you can afford your project cards, which you have to buy before Prelude benefits, so you might still get an error when you try to start. That may be a little odd but I think it doesn't represent a real problem. The remedy would be to show *both* pre-Preludes and post-Preludes values, but that's just too much information 99.99% of the time.)

It *could* display a number even with no Corp selected, which would probably be negative if you select your project cards first, but I agree that there's not much value in this, and it seems cleaner to just not show it yet. Incidentally, when deciding whether to show this number, please check whether the number of Corporations is *exactly* one. That way it will function properly whether the request in bug 1400 ever happens or not. I see no need to check that Preludes==2 though.

**Comment by kevinb9n** [comment]
Pharmacy Union makes a good example of the importance of taking corp reveal effects into account. That player would be unhappy to discover they have 8 less money available than they thought.

**Comment by kevinb9n** [comment]
Well, it finally happened. Had a beautiful thing set up with Phobolog, felt great about it, didn't even need any cash for any of my cards, and at the last minute was like "no I'll switch this prelude to Business Empire instead" and made an off by one math error in my head. Hit the button and discover that Business Empire is unplayable. What a heartbreaking feeling! Thankfully it was a solo game and I could just trash it, but it would have been torture to have to play this whole game out.

Soo I wanted to know if there's still interest in doing this!

**Comment by bafolts** [comment]
@pierrehilbert you can assign this to me if you aren't planning on working on it.

**Comment by vsrisuknimit** [comment]
@bafolts @pierrehilbert any update?

**Comment by vsrisuknimit** [comment]
Alright I want to work on this. Could someone give me suggestions to start me out. This will likely be too big for me. @bafolts ?

I can access corporation.startingMegaCredits from Game.ts. I don't know how to access the selected preludes or how to display the value in PlayerHome.ts.

I want to show three lines.
(1) Starting mega credit after corporation reveal effect (+4 mc for Splice, +3 mc for Vitor, -8 mc for Pharmacy Union).
(2) How much mc is left after keeping x project cards. (Applies effect from Terralabs and Polyphemos)
(3) How much mc after playing preludes. (Applies effect for Manutech, Tharsis with city preludes)

Error if 2 is negative. Warning if 3 is negative but still allows to play.

**Comment by kberg** [comment]
Suggestion on how to start: Well, what are you already familiar with, and what do you need for direction?

A tip: when I finally got to having to change the view to show ares global parameters, i started with just a <div> full of text that got updated as the game progressed. Only after launch did I bother to assemble something much more sophisticated. Solving one problem at a time would be great.

So if you just wanted to make ONE change, make it. Draw it in a div. Share the results with me and I'll give you feedback on where to go next?

Or, send it out for review with a screenshot and a `v-if:false` since it's not ready to show.

**Comment by vsrisuknimit** [comment]
PlayerHome handles the graphics. When there is no corp, it runs waiting-for.

Something happen in between. Then gotoInitialResearchPhase calls pickCorporationCard which then calls on SelectCard for Corps, Preludes and Projects. The calculation is done there whether a player is trying to buy too many cards.

@kberg I don't know where to put the div. Probably PlayerHome but I don't see how it will get all the data it needs.

**Comment by kberg** [comment]
@vsrisuknimit I'd put it in `SelectCard.ts`. It gets complicated because `SelectCard.ts` has a (smartly) hard-coded min and max, and no room for cost-per-card, and megacredits.

(Probably gets even more complicated because Helion makes you pay with heat, but that's a sophistication you can add later.)

**Comment by vsrisuknimit** [comment]
@kberg I can't do that. pickCorporationCard calls SelectCard 3 times for Corps, Preludes and Projects. They have to talk to each other somehow. So one div each in SelectCard does not make sense.

**Comment by kberg** [comment]
Can't do what? Change SelectCard? I think you can. Since you asked for a suggestion on where to start, that's where I'd suggest starting. See if that works or if you can find a better option elsewhere -- sorry -- I have to do my day job now! DM me if you have a couple of questions.

**Comment by chosta** [comment]
Making some progress on this. Will keep you updated.

**Comment by chosta** [comment]
I need advice for how to approach this. I found two ways of doing it but first let me explain where the complication comes from. In order to display some calculation of remaining MC upon card selection, we need to be able to access the data that is being manipulated in multiple Vue components. In this case it's an `AndOptions` that has a bunch of `SelectCard` inputs. The `playerinput` property should be sufficient to get all the data we need (it changes upon card selection). The issue comes with the way Vue works. If I add a component that displays the MC being changed upon card creation (I've called it `McCalculator`), the data needs to be updated by event listeners that are inside multiple components  (the 3x `SelectCard` ones). There are two ways to do that. Use the `data` property of `AndOptions` and pass functions that update this data down the chain. OR use global Vue state (the thing that is in App). I also noticed that `playerinput` as a prop sent to the component (McCalc) in the AndOptions does change but not sure how to access it properly.

basically in AndOptions I do this:
```
...
children.push(
      createElement(McCalculator, {
        props: {
          data: playerInput,
        },
      }),
    );
    if (this.showsave) {
...

And the `McCalculator` should be able to use the playerinput data to calculate the amount BUT I am a bit stuck at the moment. I can send a function to every SelectCard element that changes the AndOptions state (data) and I am pretty sure it would work just thinking I might be able to use the `playerinput` directly if possible.
Anyway, if this does not make sense to any of you, I'll make a draft PR and we can discuss the code step by step. Also, I don't like the 'global state' approach, its an anti-pattern.

**Comment by chosta** [comment]
I'm actually going to back of this issue for now. I thought it would be more trivial and I want to focus on other tasks in my backlog. I suggest @bafolts takes a look and make a suggestion. Thanks.

**Comment by bafolts** [comment]
I can check this out. I agree that using something with global state is not ideal. Looking at the problem at hand using events seems like it would make sense. Assuming vue can bubble up custom events for us out from `SelectCard` through the `AndOption`. We need to know when `SelectCard` has changed which cards it selected. I have no idea if that would work but I see that vue has something with custom events.

https://vuejs.org/v2/guide/components-custom-events.html

All of the player inputs are server driven for the most part. Modifying the playerinput could work but would feel wrong since that data now is what we received from the server unaltered. Starting to alter it could get dangerous quickly. Doing this some way which avoids using global state should make other things possible on other pages.

I can take a shot at this.

**Comment by chosta** [comment]
@bafolts 
It does work as expected but for me its a bit too much to do (on top of an already abstracted code that is AndOptions).
In React you usually approach this by having a parent element's state propagate to the children as `props`. In order for a child element to be able to change the parent's state, you should also propagate functions that change the parent's state and are defined in the parent. It's a pretty common pattern and It did work here too. I passed an `increment` function and was able to change the parent's state and display it in the child element. The function was invoked with :onClick in the child element.
in `AndOptions.ts`
```
...
data: function () {
   return {
      count: 0,
   }
}
...
children.push(
      createElement(McCalculator, {
        props: {
          count: this.count,
          increment: this.increment,
        },
      }),
    );
    if (this.showsave) {
...
...
methods: {
   increment: function() {
      this.count++;
   }
}
...
```
and in `McCalculator.ts`
```
<span @click="increment"/>
<span>{{count}}</span>
```
